Fix documentation in `lib/views/main_screen.dart`

**Describe the bug**
1. Our push workflow [Add tests for UserTasksPage (#1468)](https://github.com/PalisadoesFoundation/talawa/actions/runs/4116213093) is failing.
2. This is most likely due to recent changes to `lib/views/main_screen.dart`
3. This has probably never been noticed before as we have not been adding new features to the app since we implemented the documentation workflow.

**To Reproduce**
Steps to reproduce the behavior:

1. Push code to repo

**Expected behavior**
1. No error
2. A way to detect these types of errors as part of the PR process. Possibly adapting the `documentationcheck.py` script

**Actual behavior**
Error:

```
Run git branch
* develop
Defaulting to user installation because normal site-packages is not writeable
Collecting GitPython
  Downloading GitPython-3.1.30-py3-none-any.whl (184 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 184.0/184.0 KB 4.9 MB/s eta 0:00:00
Collecting gitdb<5,>=4.0.1
  Downloading gitdb-4.0.10-py3-none-any.whl (62 kB)
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 62.7/62.7 KB 11.8 MB/s eta 0:00:00
Collecting smmap<6,>=3.0.1
  Downloading smmap-5.0.0-py3-none-any.whl (24 kB)
Installing collected packages: smmap, gitdb, GitPython
Successfully installed GitPython-3.1.30 gitdb-4.0.10 smmap-5.0.0
🔍 DOCUMENTATION NOT UPDATED: Files with missing or not updated documentation found
>>> File name: lib/views/main_screen.dart
	DocumentationStatus.not_updated

Error: Process completed with exit code 1.
```

**Screenshots**
If applicable, add screenshots to help explain your problem.
![image](https://user-images.githubusercontent.com/16875803/217312125-1cd7d969-9561-4e11-aedb-652049e04e4e.png)

**Additional details**
Add any other context or screenshots about the feature request here.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

--- a/lib/view_model/main_screen_view_model.dart
+++ b/lib/view_model/main_screen_view_model.dart
@@ -1,8 +1,18 @@
+import 'package:flutter/foundation.dart';
 import 'package:flutter/material.dart';
+import 'package:hive/hive.dart';
 import 'package:talawa/locator.dart';
+import 'package:talawa/plugins/fetch_plugin_list.dart';
 import 'package:talawa/services/size_config.dart';
+import 'package:talawa/utils/app_localization.dart';
 import 'package:talawa/view_model/base_view_model.dart';
+import 'package:talawa/views/after_auth_screens/add_post_page.dart';
+import 'package:talawa/views/after_auth_screens/chat/chat_list_screen.dart';
+import 'package:talawa/views/after_auth_screens/events/explore_events.dart';
+import 'package:talawa/views/after_auth_screens/feed/organization_feed.dart';
+import 'package:talawa/views/after_auth_screens/profile/profile_page.dart';
 import 'package:talawa/widgets/custom_alert_dialog.dart';
+import 'package:talawa/widgets/theme_switch.dart';
 import 'package:tutorial_coach_mark/tutorial_coach_mark.dart';
 
 /// MainScreenViewModel class provide methods to interact with the modal to
@@ -68,20 +78,21 @@ class MainScreenViewModel extends BaseModel {
   late TutorialCoachMark tutorialCoachMark;
   final List<TargetFocus> targets = [];
 
-  int currentIndex = 0;
-  onTabTapped(int index) {
-    currentIndex = index;
-    notifyListeners();
-  }
-
-  // initialiser.
-  initialise(
+  void initialise(
     BuildContext ctx, {
     required bool fromSignUp,
     required int mainScreenIndex,
   }) {
-    currentIndex = mainScreenIndex;
+    currentPageIndex = mainScreenIndex;
     showAppTour = fromSignUp;
+
+    pluginPrototypeData = {
+      "Donation": {
+        "icon": Icons.attach_money_outlined,
+        "page": const ChangeThemeTile(),
+      }
+    };
+
     notifyListeners();
     if (!showAppTour) {
       tourComplete = true;
@@ -109,6 +120,131 @@ class MainScreenViewModel extends BaseModel {
         ),
       );
     }
+  }
+
+  /// Contains the Widgets to be rendered for corresponding navbar items.
+  /// Features that should be implemented as plugins should be kept here.
+  List<StatelessWidget> pages = [];
+
+  /// Actual [BottomNavigationBarItem]s that show up on the screen
+  List<BottomNavigationBarItem> navBarItems = [];
+
+  /// Maps the feature names with their proper Icon and Page.
+  /// `icon` contains the [IconData] corresponding to plugin's icon.
+  /// `page` contains the corresponding page to be displayed
+  /// Name of the feature provided by the admin must [exactly] match with the
+  /// name stored here.
+  Map<dynamic, dynamic> pluginPrototypeData = {};
+
+  /// Contains plugin data fetched from the server
+  List<dynamic> pluginList = [];
+
+  /// Dynamically adds [BottomNavigationBarItems] in `BottomNavigationBar`
+  /// by mapping over the data received from the server.
+  void fetchAndAddPlugins(
+    BuildContext context,
+  ) {
+    navBarItems = [
+      BottomNavigationBarItem(
+        icon: Icon(
+          Icons.home,
+          key: keyBNHome,
+        ),
+        label: AppLocalizations.of(context)!.strictTranslate('Home'),
+      ),
+      BottomNavigationBarItem(
+        icon: Icon(
+          Icons.event_note,
+          key: keyBNEvents,
+        ),
+        label: AppLocalizations.of(context)!.strictTranslate('Events'),
+      ),
+      BottomNavigationBarItem(
+        icon: Icon(
+          Icons.add_box,
+          key: keyBNPost,
+        ),
+        label: AppLocalizations.of(context)!.strictTranslate('Add'),
+      ),
+      BottomNavigationBarItem(
+        icon: Icon(
+          Icons.chat_outlined,
+          key: keyBNChat,
+        ),
+        label: AppLocalizations.of(context)!.strictTranslate('Chat'),
+      ),
+      BottomNavigationBarItem(
+        icon: Icon(
+          Icons.account_circle,
+          key: keyBNProfile,
+        ),
+        label: AppLocalizations.of(context)!.strictTranslate('Profile'),
+      )
+    ];
+
+    pages = [
+      OrganizationFeed(
+        key: const Key("HomeView"),
+        homeModel: this,
+      ),
+      ExploreEvents(
+        key: const Key('ExploreEvents'),
+        homeModel: this,
+      ),
+      AddPost(
+        key: const Key('AddPost'),
+        drawerKey: MainScreenViewModel.scaffoldKey,
+      ),
+      const ChatPage(
+        key: Key('Chats'),
+      ),
+      ProfilePage(
+        key: keySPEditProfile,
+        homeModel: this,
+      ),
+    ];
+
+    pluginList = (Hive.box('pluginBox').get('plugins') ?? []) as List<dynamic>;
+
+    pluginList.forEach((plugin) {
+      if (pluginPrototypeData.containsKey(plugin["pluginName"] as String) &&
+          plugin["pluginInstallStatus"] as bool) {
+        navBarItems.add(
+          BottomNavigationBarItem(
+            icon: Icon(
+              pluginPrototypeData[plugin["pluginName"]]["icon"] as IconData,
+            ),
+            label: AppLocalizations.of(context)!.strictTranslate(
+              plugin["pluginName"] as String,
+            ),
+          ),
+        );
+        pages.add(
+          pluginPrototypeData[plugin["pluginName"]]["class"] as StatelessWidget,
+        );
+      }
+    });
+
+    /// Causes the app to continously check for plugins if they are
+    /// updated and re-render the navbar
+    WidgetsBinding.instance.addPostFrameCallback((_) {
+      FetchPluginList();
+      final newPluginList =
+          (Hive.box('pluginBox').get('plugins') ?? []) as List<dynamic>;
+
+      if (listEquals(pluginList, newPluginList)) {
+        notifyListeners();
+      }
+    });
+  }
+
+  /// Currently selected page
+  int currentPageIndex = 0;
+
+  /// Handles click on [BottomNavigationBarItem]
+  void onTabTapped(int index) {
+    currentPageIndex = index;
+    notifyListeners();
   }
 
   /// This function show tutorial to user.
@@ -217,7 +353,7 @@ class MainScreenViewModel extends BaseModel {
     showTutorial(
       onClickTarget: showHome,
       onFinish: () {
-        onTabTapped(currentIndex + 1);
+        onTabTapped(currentPageIndex + 1);
         if (!tourComplete && !tourSkipped) {
           tourEventTargets();
         }
@@ -279,7 +415,7 @@ class MainScreenViewModel extends BaseModel {
     );
     showTutorial(
       onFinish: () {
-        onTabTapped(currentIndex + 1);
+        onTabTapped(currentPageIndex + 1);
         if (!tourComplete && !tourSkipped) {
           tourAddPost();
         }
@@ -302,7 +438,7 @@ class MainScreenViewModel extends BaseModel {
     );
     showTutorial(
       onFinish: () {
-        onTabTapped(currentIndex + 1);
+        onTabTapped(currentPageIndex + 1);
         if (!tourComplete && !tourSkipped) {
           tourChat();
         }
@@ -325,7 +461,7 @@ class MainScreenViewModel extends BaseModel {
     );
     showTutorial(
       onFinish: () {
-        onTabTapped(currentIndex + 1);
+        onTabTapped(currentPageIndex + 1);
         if (!tourComplete && !tourSkipped) {
           tourProfile();
         }

--- a/lib/views/main_screen.dart
+++ b/lib/views/main_screen.dart
@@ -1,20 +1,8 @@
 import 'package:flutter/material.dart';
-import 'package:hive/hive.dart';
 import 'package:talawa/models/mainscreen_navigation_args.dart';
-import 'package:talawa/plugins/fetch_plugin_list.dart';
-import 'package:talawa/utils/app_localization.dart';
 import 'package:talawa/view_model/main_screen_view_model.dart';
-import 'package:talawa/views/after_auth_screens/add_post_page.dart';
-import 'package:talawa/views/after_auth_screens/chat/chat_list_screen.dart';
-import 'package:talawa/views/after_auth_screens/events/explore_events.dart';
-import 'package:talawa/views/after_auth_screens/feed/organization_feed.dart';
-import 'package:talawa/views/after_auth_screens/profile/profile_page.dart';
 import 'package:talawa/views/base_view.dart';
 import 'package:talawa/widgets/custom_drawer.dart';
-import 'package:talawa/widgets/theme_switch.dart';
-
-late List<StatelessWidget> navBarClasses;
-late List<BottomNavigationBarItem> navBarItems;
 
 class MainScreen extends StatefulWidget {
   const MainScreen({Key? key, required this.mainScreenArgs}) : super(key: key);
@@ -25,132 +13,8 @@ class MainScreen extends StatefulWidget {
 }
 
 class _MainScreenState extends State<MainScreen> {
-  final Map<String, bool> addedToNav = {
-    "Chats": true,
-    "Posts": true,
-    "Events": true,
-  };
-
-  List<List<Object>> renderBottomNavBarPlugins(
-      List<BottomNavigationBarItem> navBarItems,
-      List<StatelessWidget> navBarClasses,
-      Map navNameClasses,
-      Map<String, Widget> navNameIcons,
-      BuildContext context) {
-    //get list from hive
-    final Box box;
-    // ignore: prefer_typing_uninitialized_variables
-    var pluginList;
-    const bool res = false;
-    box = Hive.box('pluginBox');
-    pluginList = box.get('plugins');
-
-    /// CAUTION : `pluginList` is currently filled with dummy data in order to make sure that
-    /// all features of talawa should be visible from the navbar.
-    /// following dummy data should be only used for developement purpose.
-    /// Adding dummy data enables to use Events,Posts,Chats features event if plugin documents for them is not created by the admin.
-    /// FOR PRODUCTION : please make this array empty to enable or test real plugin data from the server
-    pluginList ??= [
-      // if null then it will be replace with the following data
-      {
-        "_id": "62cfcd6e33bbe266f59644dd",
-        "pluginInstallStatus": true,
-        "pluginName": "Events",
-        "pluginCreatedBy": " Mr. Siddhesh Bhupendra  Kukade",
-        "pluginDesc":
-            "Presenting the Events Plugin by Siddhesh Bhupendra Kuakde that implements the plugin features as a component here members can create and manage events and on top of it everything is controlled by the Organization Admins . Install and use right away.",
-        "installedOrgs": ["<YOUR_ORG_ID_FOR_DEVELOPMENT>"]
-      },
-      {
-        "_id": "62cfcd6233bbe266f59644db",
-        "pluginInstallStatus": true,
-        "pluginName": "Posts",
-        "pluginCreatedBy": "Mr.Johnathan Doe ",
-        "pluginDesc":
-            "Presenting the Donation plugin for the talawa app that enables members of the organization to  do One time or Reccurinng donations to an  organization",
-        "installedOrgs": ["<YOUR_ORG_ID_FOR_DEVELOPMENT>"]
-      },
-      {
-        "_id": "62cfcf71824dc26e1fbeed46",
-        "pluginInstallStatus": true,
-        "pluginName": "Donation",
-        "pluginCreatedBy": "Kimia",
-        "pluginDesc":
-            "See all available organizations in entire world using this plugins",
-        "installedOrgs": ["<YOUR_ORG_ID_FOR_DEVELOPMENT>"]
-      },
-      {
-        "_id": "62cfcf71824dc26e1fbeed46",
-        "pluginInstallStatus": true,
-        "pluginName": "Chats",
-        "pluginCreatedBy": "Kimia",
-        "pluginDesc":
-            "See all available organizations in entire world using this plugins",
-        "installedOrgs": ["<YOUR_ORG_ID_FOR_DEVELOPMENT>"]
-      }
-    ];
-
-    /// Here we are dynamically adding BottomNavigationBarItems in navbar
-    /// by mapping over the data received from the server.
-    pluginList.forEach((plugin) {
-      if (navNameClasses.containsKey(plugin["pluginName"] as String) &&
-          plugin["pluginInstallStatus"] as bool &&
-          !addedToNav.containsKey(plugin["pluginName"])) {
-        navBarItems.add(
-          BottomNavigationBarItem(
-            icon: navNameIcons[plugin["pluginName"]] ?? const Icon(Icons.home),
-            label: AppLocalizations.of(context)!
-                .strictTranslate(plugin["pluginName"] as String),
-          ),
-        );
-        navBarClasses.add(
-          navNameClasses[plugin["pluginName"]]["class"] as StatelessWidget,
-        );
-        // addedToNav[plugin["pluginName"] as String] = true;
-        addedToNav.putIfAbsent(
-          plugin["pluginName"] as String,
-          () => true,
-        );
-      }
-    });
-
-    /// updating the state to re-render the navbar widget.
-    WidgetsBinding.instance.addPostFrameCallback((_) {
-      /// This line causes the app to continiously check for plugins if they are updated
-      /// and it will automatically re-render the navbar with enables or disabled features
-      FetchPluginList(); // dont delete
-      setState(() {});
-    });
-
-    return [navBarItems, navBarClasses];
-  }
-
   @override
   Widget build(BuildContext context) {
-    ///Features that are not meant to be implemented as plugins should be kept here.
-    navBarItems = [
-      BottomNavigationBarItem(
-        icon: const Icon(Icons.home),
-        label: AppLocalizations.of(context)!.strictTranslate('Home'),
-      ),
-      BottomNavigationBarItem(
-        icon: const Icon(Icons.event_note),
-        label: AppLocalizations.of(context)!.strictTranslate('Events'),
-      ),
-      BottomNavigationBarItem(
-        icon: const Icon(Icons.add_box),
-        label: AppLocalizations.of(context)!.strictTranslate('Add'),
-      ),
-      BottomNavigationBarItem(
-        icon: const Icon(Icons.chat_outlined),
-        label: AppLocalizations.of(context)!.strictTranslate('Chat'),
-      ),
-      BottomNavigationBarItem(
-        icon: const Icon(Icons.account_circle),
-        label: AppLocalizations.of(context)!.strictTranslate('Profile'),
-      )
-    ];
-
     return BaseView<MainScreenViewModel>(
       onModelReady: (model) {
         model.initialise(
@@ -160,74 +24,9 @@ class _MainScreenState extends State<MainScreen> {
         );
       },
       builder: (context, model, child) {
-        /// `navNameIcon` Object maps Icons to proper features according to their names used in navbar.
-        /// CAUTION : Name of the feature in talwa app must match with the name that is provided by the admin.
-        final Map<String, Widget> navNameIcon = {
-          "Events": const Icon(Icons.event_note),
-          "Posts": const Icon(Icons.add_box),
-          "Profile": const Icon(Icons.account_circle),
-          "Chats": const Icon(Icons.chat_outlined),
-          "Donation": const Icon(Icons.attach_money_outlined),
-        };
-
-        /// `navNameClasses` Object maps the feature names with thier proper Icons and Widgets (named as `class`) used in navbar
-        /// CAUTION : Name of the feature in talwa app must match with the name that is provided by the admin.
-        final Map<dynamic, dynamic> navNameClasses = {
-          "Events": {
-            // "icon": Icons.event_note,
-            "class": ExploreEvents(
-              key: const Key('ExploreEvents'),
-              homeModel: model,
-            )
-          },
-          "Posts": {
-            "class": AddPost(
-              key: const Key('AddPost'),
-              drawerKey: MainScreenViewModel.scaffoldKey,
-            )
-          },
-          "Chats": {
-            "class": const ChatPage(
-              key: Key('Chats'),
-            ),
-          },
-          "Donation": {
-            "class": const ChangeThemeTile(),
-          }
-        };
-
-        /// `navBarClasses` is the array that contains the Widgets to be rendered on the navbar
-        /// Features that should be implemented as plugins like Home, Profile,etc. should be kept here
-        /// When talawa app receives the plugins data is will dynamically render more components from the construtor.
-        navBarClasses = [
-          OrganizationFeed(
-            key: const Key("HomeView"),
-            homeModel: model,
-          ),
-          ExploreEvents(
-            key: const Key('ExploreEvents'),
-            homeModel: model,
-          ),
-          AddPost(
-            key: const Key('AddPost'),
-            drawerKey: MainScreenViewModel.scaffoldKey,
-          ),
-          const ChatPage(
-            key: Key('Chats'),
-          ),
-          ProfilePage(
-            key: model.keySPEditProfile,
-            homeModel: model,
-          ),
-        ];
-        // ignore_for_file:  unused_local_variable
-        final res = renderBottomNavBarPlugins(
-          navBarItems,
-          navBarClasses,
-          navNameClasses,
-          navNameIcon,
-          context,
-        );
+        // Checks for updates in plugins from the server.
+        // Will continously hit the server and fetch plugin information.
+        model.fetchAndAddPlugins(context);
 
         return Scaffold(
           key: MainScreenViewModel.scaffoldKey,
@@ -236,46 +35,18 @@ class _MainScreenState extends State<MainScreen> {
             key: const Key("Custom Drawer"),
           ),
           body: IndexedStack(
-            index: model.currentIndex,
-            children: navBarClasses,
+            index: model.currentPageIndex,
+            children: model.pages,
           ),
-          bottomNavigationBar: Stack(
-            children: [
-              SizedBox(
-                height: 50,
-                child: Row(
-                  children: [
-                    getBNBItem(model.keyBNHome),
-                    getBNBItem(model.keyBNEvents),
-                    getBNBItem(model.keyBNPost),
-                    getBNBItem(model.keyBNChat),
-                    getBNBItem(model.keyBNProfile),
-                  ],
-                ),
-              ),
-              BottomNavigationBar(
-                type: BottomNavigationBarType.fixed,
-                currentIndex: model.currentIndex,
-                onTap: model.onTabTapped,
-                selectedItemColor: const Color(0xff34AD64),
-                items: navBarItems,
-              ),
-            ],
+          bottomNavigationBar: BottomNavigationBar(
+            type: BottomNavigationBarType.fixed,
+            currentIndex: model.currentPageIndex,
+            onTap: model.onTabTapped,
+            selectedItemColor: const Color(0xff34AD64),
+            items: model.navBarItems,
           ),
         );
       },
-    );
-  }
-
-  Widget getBNBItem(GlobalKey<State<StatefulWidget>> key) {
-    return Expanded(
-      child: Center(
-        child: SizedBox(
-          key: key,
-          height: 40,
-          width: 40,
-        ),
-      ),
     );
   }
 }

--- a/test/view_model_tests/main_screen_view_model_test.dart
+++ b/test/view_model_tests/main_screen_view_model_test.dart
@@ -48,7 +48,7 @@ void main() {
 
     test("When initialized current index should be 0", () {
       final mainTestModel = getModel();
-      expect(mainTestModel.currentIndex, 0);
+      expect(mainTestModel.currentPageIndex, 0);
     });
   });
 
@@ -58,7 +58,7 @@ void main() {
         () {
       final mainTestModel = getModel();
       mainTestModel.onTabTapped(4);
-      expect(mainTestModel.currentIndex, 4);
+      expect(mainTestModel.currentPageIndex, 4);
     });
 
     test("When called function should notify listeners of tab change", () {
@@ -99,7 +99,7 @@ void main() {
           fSignUp: fSignup, mainIndex: mainIndex, model: mainTestModel);
 
       expect(mainTestModel.showAppTour, fSignup);
-      expect(mainTestModel.currentIndex, mainIndex);
+      expect(mainTestModel.currentPageIndex, mainIndex);
     });
 
     test(

--- a/test/views/main_screen_test.dart
+++ b/test/views/main_screen_test.dart
@@ -151,7 +151,7 @@ void main() async {
               widget is Scaffold &&
               widget.drawer is CustomDrawer &&
               widget.body is IndexedStack &&
-              widget.bottomNavigationBar is Stack,
+              widget.bottomNavigationBar is BottomNavigationBar,
         ),
         findsOneWidget,
       );

--- a/test/views/main_screen_test.dart
+++ b/test/views/main_screen_test.dart
@@ -26,7 +26,6 @@ import 'package:talawa/views/after_auth_screens/profile/profile_page.dart';
 import 'package:talawa/views/base_view.dart';
 import 'package:talawa/views/main_screen.dart';
 import 'package:talawa/widgets/custom_drawer.dart';
-import 'package:talawa/widgets/theme_switch.dart';
 
 import '../helpers/test_helpers.dart';
 import '../helpers/test_locator.dart';
@@ -185,10 +184,9 @@ void main() async {
         expect(find.byType(AddPost), findsOneWidget);
         expect(find.byType(ProfilePage), findsOneWidget);
 
-        // If MainScreen finds the saved plugin to be nil,
-        // it will add some by itself
+        // If MainScreen finds some plugins, it will add them dynamically
 
-        expect(find.byType(ChangeThemeTile), findsOneWidget);
+        // expect(find.byType(ChangeThemeTile), findsOneWidget);
       });
     });
 
@@ -199,7 +197,7 @@ void main() async {
 
         await tester.tap(find.byIcon(Icons.home));
         await tester.pump();
-        expect(mainScreenViewModel.currentIndex, 0);
+        expect(mainScreenViewModel.currentPageIndex, 0);
 
         await tester.tap(find.byIcon(Icons.event_note));
         await tester.pump();

